### ISSUE TITLE
OpenQASM returned by convert_to_openqasm includes opaque gates.

### ISSUE THREAD

**Issue by hodgestar** [sandialabs/pyGSTi#377]
**Describe the bug**

PyGSTI 0.9.12 introduce a new `opaque delay(t) q;` gate definition into the OpenQASM 2 produced by `convert_to_openqasm` in commit 4b740b3f1cd67ce6f0aab37fc77eb26b9699fe48. This defines an unspecified opaque gate named `delay` making the meaning of the returned OpenQASM program undefined.

Could this be replaced with a defined OpenQASM 2 construct?

It's not entirely clear to me why the old `id` gate was there either. If we want to ensure that all gates are used in portion of the output program, wouldn't a barrier on all the qubits suffice?

Perhaps it is also time to add an OpenQASM 3 exporter? If that's desired I could make an attempt and create a merge request.

Link to [OpenQASM 2 spec](https://arxiv.org/pdf/1707.03429v2.pdf) so that I don't have to search for it again later.

**To Reproduce**

Run `convert_to_openqasm`.

**Expected behavior**

The returned OpenQASM program does not contain undefined gates.

**Environment (please complete the following information):**
 - pyGSTi version: v0.9.12

**Comment by hodgestar** [comment]
@sserita Thanks for the quick reply. Very happy for this to be a discussion instead of a bug.

For adding OpenQASM 3 support, my thought was to just add a `convert_to_openqasm3` method. The `convert_to_openqasm` method could then be deprecated at some point down the line. Alternatively, one could add a version parameter to `convert_to_openqasm`. Somehow the separate method seems nicer to me, but I'd be happy to write it either way.

I'm still curious about why these zero delays or id gates are needed here.

**Comment by sserita** [comment]
Re: why they are needed: That's a fair question. It's almost certainly for some legacy reason - I will check with other team members and see if we remember why these were added and if we can remove them. That way, we could potentially fix this issue for you in the short-term, and spawn off the QASM3 converter as a new feature.

I can see both methods working, but I'm inclined to agree that a separate function seems a little cleaner. That way, we can just use a deprecated decorator on the old function whenever we want to remove it (which may not be quite yet, as there is still a lot of QASM2 floating around).

We would still need a Contributor License Agreement (CLA) from you before we could accept a PR - email me at sserita@sandia.gov to get that process started.

**Comment by hodgestar** [comment]
@sserita Any luck finding out why these strange operations are needed?

**Comment by sserita** [comment]
Hey @hodgestar, there's now a flag to remove those delay instructions. You can pass `include_delay_on_idle=False` into `convert_to_openqasm` and it should drop the delays and be a valid OpenQASM. It's currently just in `bugfix-release-0.9.12.1` branch, but that release will be out shortly.

I'm going to close this out if that's acceptable, and we can create a new issue whenever we start working on the OpenQASM 3.0 emitter.

**Comment by hodgestar** [comment]
Thanks @sserita. I was waiting on hearing what the delays were for before starting an OpenQASM 3 PR, but I'll take the new flag as meaning that the delays for zero time are probably not necessary.

Happy with this issue being closed. I'll open a PR for the OpenQASM 3 variant when I get there.

**Comment by sserita** [comment]
Yep that sounds good! In OpenQASM 3, we actually have delay as a valid instruction (and stretches to deal with variable time of the other gates so we don't have to hardcode a 0), but I can handle that logic when we get there if that's something that doesn't interest you in implementing. 😅

**Comment by hodgestar** [comment]
I'll figure it out -- you can comment on the PR if I get it wrong. :)
